src/popover/__tests__/popover.test.tsx

```diff
@@ -415,3 +421,90 @@ test('does not add portal to the body unless visible', () => {
+  // instances corresponding to those nodes, and manually call onClick on them. See
+  // https://github.com/facebook/react/issues/10422#issuecomment-674928774
+  function retargetClicks(root: HTMLElement) {
+    function findReactComponent(item: EventTarget) {
+      for (const key in item) {
+        if (Object.prototype.hasOwnProperty.call(item, key) && key.indexOf('_reactInternal') !== -1) {
+          return item[key as keyof typeof item] as any;
+        }
+      }
+    }
+
+    root.addEventListener('click', event => {
+      const eventPath = event.composedPath();
+      for (const i in eventPath) {
+        const item = eventPath[i];
+
+        const internalComponent = findReactComponent(item);
+        if (internalComponent && internalComponent.memoizedProps) {
+          internalComponent.memoizedProps.onClick?.(event);
+          break;
+        }
+
+        if (item === root) {
+          break;
+        }
+      }
+    });
+  }
+
+  function renderPopoverInShadowDom(props: PopoverProps & { ref?: React.Ref<PopoverProps.Ref> }) {
+    const shadowContainer = document.createElement('div');
+    shadowContainer.setAttribute('id', 'shadow-container');
+    shadowContainer.attachShadow({ mode: 'open' });
+
+    const shadowChild = document.createElement('div');
+    shadowChild.setAttribute('id', 'shadow-child');
+    shadowContainer.shadowRoot!.appendChild(shadowChild);
+
+    document.body.appendChild(shadowContainer);
+
+    retargetClicks(shadowChild);
+
+    const { container } = render(<Popover {...props} />, { container: shadowChild });
+
+    return new PopoverInternalWrapper(container);
+  }
+
+  afterEach(() => {
+    // cleanup test container after each test
+    document.getElementById('shadow-container')?.remove();
+  });
+
+  it('renders text trigger correctly', () => {
+    const wrapper = renderPopoverInShadowDom({ children: 'Trigger', triggerAriaLabel: 'Test aria label' });
+    expect(wrapper.findTrigger().getElement().tagName).toBe('BUTTON');
+    expect(wrapper.findTrigger().getElement()).toHaveTextContent('Trigger');
+    expect(wrapper.findTrigger().getElement()).toHaveAccessibleName('Test aria label');
+  });
+
+  it('renders the popover header correctly', () => {
+    const wrapper = renderPopoverInShadowDom({ children: 'Trigger', header: 'Memory error' });
+    wrapper.findTrigger().click();
+
+    expect(wrapper.findHeader()!.getElement()).toHaveTextContent('Memory error');
+    // this assertion fails if Popover bails early on calculating its position
+    expect(wrapper.findContainer()!.getElement()).toHaveAttribute(
+      'style',
+      'inset-block-start: 0; inset-inline-start: 13px;'
+    );
+  });
+
+  it('renders the popover header correctly with portal', () => {
+    const wrapper = renderPopoverInShadowDom({ children: 'Trigger', header: 'Memory error', renderWithPortal: true });
+    wrapper.findTrigger().click();
```

Comment:
Considering the complications of supporting the click() event, how about we scope the tests to the popover container (internal) component instead? You can add a new test file `popover-container.test.tsx` and assert the expected placement, but without the need to make it visible first.